File: Leetcode 0297 Serialize and Deserialize Binary Tree [hard] [tree]/Serialize and Deserialize Binary Tree [bfs].java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) {
            return "";
        }
        
        Queue<TreeNode> queue = new LinkedList<>();
        StringBuilder result = new StringBuilder();
        
        queue.offer(root);
        while (!queue.isEmpty()) {
            TreeNode cur = queue.poll();
            if (cur == null) {
                result.append("null");
            } else {
                result.append(cur.val);
                queue.offer(cur.left);
                queue.offer(cur.right);
            }
            result.append(",");
        }
        
        return result.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.equals("")) {
            return null;
        }
        
        String[] result = data.split(",");
        Queue<TreeNode> queue = new LinkedList<>();
        TreeNode root = new TreeNode(Integer.parseInt(result[0]));
        queue.offer(root);
        
        for (int i = 1; i < result.length; i++) {
            TreeNode cur = queue.poll();
            if (!result[i].equals("null")) {
                TreeNode leftChild = new TreeNode(Integer.parseInt(result[i]));
                queue.offer(leftChild);
                cur.left = leftChild;
            }
            
            if (!result[++i].equals("null")) {
                TreeNode rightChild = new TreeNode(Integer.parseInt(result[i]));
                queue.offer(rightChild);
                cur.right = rightChild;
            }
        }
                                    
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: Leetcode 0297 Serialize and Deserialize Binary Tree [hard] [tree]/Serialize and Deserialize Binary Tree [bfs].js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
    if (root === null) {
        return "";
    }
    
    let result = "";
    let queue = [];
    queue.push(root);
    while (queue.length > 0) {
        let cur = queue.shift();
        if (cur !== null) {
            result += cur.val;
            queue.push(cur.left);
            queue.push(cur.right);
        } else {
            result += "null";
        }
        result += ","
    }
    console.log(result);
    
    return result;
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    if (data === "") {
        return null;
    }
    
    let queue = [];
    let result = data.split(",");
    console.log(result);
    let root = new TreeNode(result[0]);
    queue.push(root);
    
    for (let i = 1; i < result.length; i++) {
        // Handle last one ("") in result array after split
        if (result[i] === "") {
            continue;
        }
        
        let cur = queue.shift();
        if (result[i] !== "null") {
            let leftChild = new TreeNode(result[i]);
            cur.left = leftChild;
            queue.push(leftChild);
        }
        
        if (result[++i] !== "null") {
            let rightChild = new TreeNode(result[i]);
            cur.right = rightChild;
            queue.push(rightChild);
        }
    }
    
    return root;
};

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */